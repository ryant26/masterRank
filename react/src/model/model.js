import {
    addHero as addHeroAction,
    removeHero as removeHeroAction,
    clearAllHeroes as clearAllHeroesAction
} from "../actionCreators/hero";
import {
    addHero as addPreferredHeroAction,
    removeHero as removePreferredHeroAction,
    updateHeroes as updatePreferredHeroesAction
} from "../actionCreators/preferredHeroes";
import { updateUser as updateUserAction } from "../actionCreators/user";
import {
    addFilter as addFilterAction,
    removeFilter as removeFilterAction,
    removeAllFilters as removeAllFiltersAction
} from "../actionCreators/heroFilters";
import {
    updateGroup as updateGroupAction,
    leaveGroup as leaveGroupAction
} from '../actionCreators/group';
import { clientEvents } from "../api/websocket";
import {
    addGroupInvite as addGroupInviteAction,
    removeGroupInvite as removeGroupInviteAction
} from '../actionCreators/groupInvites';
import {
    pushBlockingEvent as pushBlockingLoadingAction,
    popBlockingEvent as popBlockingLoadingAction,
} from "../actionCreators/loading";

import {
    joinedGroupNotification,
    userJoinedGroupNotification,
    inviteSentNotification,
    inviteReceivedNotification,
    successfullyLeftGroupNotification,
    errorNotification
} from '../components/Notifications/Notifications';

import NotRealHeroes from '../resources/metaListFillerHeroes';

let socket;
let store;

const initialize = function(passedSocket, passedStore) {
    store = passedStore;
    socket = passedSocket;

    store.dispatch(pushBlockingLoadingAction());

    socket.on(clientEvents.initialData, (heroesFromServer) => _handleInitialData(heroesFromServer));
    socket.on(clientEvents.heroAdded, (hero) => _addHeroToStore(hero));
    socket.on(clientEvents.heroRemoved, (hero) => _removeHeroFromStore(hero));

    socket.on(clientEvents.groupInviteReceived, (groupInviteObject) => _addGroupInviteToStore(groupInviteObject));
    socket.on(clientEvents.playerInvited, (groupInviteObject) => _updateGroupInStore(groupInviteObject));

    socket.on(clientEvents.groupInviteCanceled, (groupInviteObject) => _removeGroupInviteFromStore(groupInviteObject));
    socket.on(clientEvents.playerInviteCanceled, (groupInviteObject) => _updateGroupInStore(groupInviteObject));

    socket.on(clientEvents.groupInviteDeclined, (groupInviteObject) => _updateGroupInStore(groupInviteObject));
    socket.on(clientEvents.groupInviteAccepted, (groupInviteObject) => _handleGroupInviteAccepted(groupInviteObject));
    socket.on(clientEvents.groupPromotedLeader, (groupInviteObject) => _updateGroupInStore(groupInviteObject));
    socket.on(clientEvents.playerHeroLeft, (groupInviteObject) => _updateGroupInStore(groupInviteObject));

    socket.on(clientEvents.error.addHero, _addHeroErrorHandler);
    socket.on(clientEvents.error.groupLeave, _groupErrorHandler);
    socket.on(clientEvents.error.groupInviteAccept, _groupErrorHandler);
    socket.on(clientEvents.error.groupInviteCancel, _groupErrorHandler);
    socket.on(clientEvents.error.groupInviteDecline, _groupErrorHandler);
};

const addHeroFilterToStore = function(filter) {
    store.dispatch(addFilterAction(filter));
};

const removeHeroFilterFromStore = function(filter) {
    store.dispatch(removeFilterAction(filter));
};

const removeAllHeroFiltersFromStore = function() {
    store.dispatch(removeAllFiltersAction());
};

const addPreferredHeroToStore = function(heroName, preference) {
    store.dispatch(addPreferredHeroAction(heroName, preference));
};

const removePreferredHeroFromStore = function(heroName, preference) {
    store.dispatch(removePreferredHeroAction(heroName, preference));
};

const updatePreferredHeroesInStore = function(heroes) {
    store.dispatch(updatePreferredHeroesAction(heroes));
};


const updatePreferredHeroes = function(heroes) {
    let currentPreferredHeroes = store.getState().preferredHeroes.heroes;
    let numberOfHeroesToCheck = Math.max(heroes.length, currentPreferredHeroes.length);

    for (let i = 0; i < numberOfHeroesToCheck; i++) {
        let currentPreferredHero = store.getState().preferredHeroes.heroes[i];
        let newPreferredHero = heroes[i];

        if (currentPreferredHero !== newPreferredHero){
            if (currentPreferredHero) {
                socket.removeHero(currentPreferredHero);
            }

            if (newPreferredHero) {
                socket.addHero(newPreferredHero, i+1);
            }
        }
    }

    updatePreferredHeroesInStore(heroes);
};

const updateUser = function(user) {
    store.dispatch(updateUserAction(user));
};

const inviteUserToGroup = function(userObject) {
    inviteSentNotification(userObject.platformDisplayName);
    socket.groupInviteSend(userObject);
};

const createNewGroup = function() {
    socket.createGroup(store.getState().preferredHeroes.heroes[0]);
};

const leaveGroup = function() {
    successfullyLeftGroupNotification(store.getState().group.leader.platformDisplayName);
    store.dispatch(leaveGroupAction());
    socket.groupLeave();
};

const cancelInvite = function(userObject) {
    socket.groupInviteCancel(userObject);
};

const acceptGroupInviteAndRemoveFromStore = function(groupInviteObject) {
    _removeGroupInviteFromStore(groupInviteObject);
    joinedGroupNotification(groupInviteObject.leader.platformDisplayName);
    socket.groupInviteAccept(groupInviteObject.groupId);
};

const declineGroupInviteAndRemoveFromStore = function(groupInviteObject) {
    _removeGroupInviteFromStore(groupInviteObject);
    socket.groupInviteDecline(groupInviteObject.groupId);
};

const loadMetaListFillerHeroes = () => {
    NotRealHeroes.forEach((hero) => {
        _addHeroToStore(hero);
    });
};

const _handleInitialData = function(heroesFromServer) {
    store.dispatch(clearAllHeroesAction());
    loadMetaListFillerHeroes();

    let userPlatformDisplayName = store.getState().user.platformDisplayName;
    heroesFromServer.forEach((hero) => {
        if(hero.platformDisplayName !== userPlatformDisplayName){
            store.dispatch(addHeroAction(hero));
        } else  {
            socket.removeHero(hero.heroName);
        }
    });

    store.getState().preferredHeroes.heroes.forEach((heroName, i) => {
         socket.addHero(heroName, (i+1));
    });

    store.dispatch(popBlockingLoadingAction());
};

const _addHeroToStore = function(hero) {
    store.dispatch(addHeroAction(hero));
    if (hero.platformDisplayName === store.getState().user.platformDisplayName) {
        addPreferredHeroToStore(hero.heroName, hero.preference);
    }

    if (store.getState().preferredHeroes.heroes[0] === hero.heroName) {
        createNewGroup(store.getState().preferredHeroes.heroes[0]);
    }
};

const _removeHeroFromStore = function(hero) {
    store.dispatch(removeHeroAction(hero));
    //TODO: Multiple tabs or window messes up preferred heroes. 
    if (hero.platformDisplayName === store.getState().user.platformDisplayName) {
        removePreferredHeroFromStore(hero.heroName, hero.priority);
    }
};

const _addHeroErrorHandler = function(error) {
    errorNotification(error.heroName);
    removePreferredHeroFromStore(error.heroName);
    // Do whatever we do to shuffle heroes in the case that multiple were added
    // After this one
};

const _addGroupInviteToStore = function(groupInviteObject) {
    inviteReceivedNotification(groupInviteObject.leader.platformDisplayName);
    store.dispatch(addGroupInviteAction(groupInviteObject));
};

const _removeGroupInviteFromStore = function(groupInviteObject) {
    store.dispatch(removeGroupInviteAction(groupInviteObject));
};

const _updateGroupInStore = function(groupInviteObject) {
    store.dispatch(updateGroupAction(groupInviteObject));
};

const _handleGroupInviteAccepted = (newGroup) => {
    //TODO: is there a better way to get the new group member?
    let previouslyPendingMembers = store.getState().group.pending;

    if (previouslyPendingMembers.length > 0 && newGroup.members.length > 0) {
        let newMember = previouslyPendingMembers.find((pendingMember) => {
            return newGroup.members.find((member) => {
                return member.platformDisplayName === pendingMember.platformDisplayName;
            });
        });
        //TODO: this if statement should never not trigger. If it does trigger that means something went wrong.
        // Should i sent a generic user joined group message or just not send a message?
        if(newMember) {
            userJoinedGroupNotification(newMember.platformDisplayName);
        }
    }

    store.dispatch(updateGroupAction(newGroup));
};

const _groupErrorHandler = (error) => {
    errorNotification(error.message);
};

const Actions = {
    initialize,
    addPreferredHeroToStore,
    updatePreferredHeroes,
    addHeroFilterToStore,
    removeHeroFilterFromStore,
    removeAllHeroFiltersFromStore,
    updateUser,
    leaveGroup,
    inviteUserToGroup,
    createNewGroup,
    cancelInvite,
    acceptGroupInviteAndRemoveFromStore,
    declineGroupInviteAndRemoveFromStore
};

export default Actions;