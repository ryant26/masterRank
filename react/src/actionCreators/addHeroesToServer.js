export function addHeroesToServer(heroNames, socket) {

    return () => {
        heroNames.forEach((heroName, i) => {
            socket.addHero(heroName, (i+1));
        });
    };
}