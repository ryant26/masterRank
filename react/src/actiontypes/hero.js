export const ADD_HERO = 'hero/ADD_HERO';
export const ADD_HEROES = 'hero/ADD_HEROES';
