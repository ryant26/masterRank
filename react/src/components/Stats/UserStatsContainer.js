import React from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import HeroStatsList from './HeroStatsList/HeroStatsList';
import HeroImages from './HeroImages/HeroImages';
import Model from '../../model/model';

const UserStatsContainer = ({hero, heroes, invitable, toggleModal}) => {
    const userHeroes = heroes.filter((userHero) => userHero.platformDisplayName === hero.platformDisplayName).sort((hero1, hero2) => {
        return hero1.priority > hero2.priority;
    });

    const wins = userHeroes.reduce((wins, hero) => {
        if (hero.stats && hero.stats.wins) {
            return wins + hero.stats.wins;
        }
        return wins;
    }, 0);

    const srNode = hero.skillRating ? <span className="sub-title"><b>{hero.skillRating}</b> SR</span> : '';
    const winsNode = wins ? <span className="sub-title"><b>{wins}</b> WINS</span> : '';

    const invitePlayer = () => {
        toggleModal();
        Model.inviteUserToGroup({
            platformDisplayName: hero.platformDisplayName,
            heroName: hero.heroName
        });
    };

    return (
        <div className="UserStatsContainer">
            <div className="header">
                <div className="flex justify-between align-center">
                    <div>
                        <h3>{hero.platformDisplayName}</h3>
                        {srNode}
                        {winsNode}
                    </div>
                    <HeroImages heroNames={userHeroes.map((hero) => hero.heroName)}/>
                </div>
            </div>
            <div className="body">
                <HeroStatsList heroes={userHeroes}/>
            </div>
            {invitable ?
                <div className="footer flex justify-end align-center">
                    <div className="button-secondary flex align-center" onClick={invitePlayer}>
                        <div className="button-content">
                            INVITE TO GROUP
                        </div>
                    </div>
                </div> :
                undefined
            }
        </div>
    );
};

UserStatsContainer.propTypes = {
    hero: PropTypes.object,
    heroes: PropTypes.array.isRequired,
    invitable: PropTypes.bool.isRequired,
    toggleModal: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
    return {
        heroes: state.heroes
    };
};

export default connect(mapStateToProps)(UserStatsContainer);