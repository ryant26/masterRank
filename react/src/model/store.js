import { createStore as createReduxStore } from 'redux';

import reducers from '../reducers/index';
import { loadState } from './reduxSerializer';


export const createStore = function() {
    store = createReduxStore(reducers, loadState());

    return store;
};

let store = createStore();

export default store;