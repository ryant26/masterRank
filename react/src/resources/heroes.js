const HEROES = [
    {
        "platformDisplayName": "PwNShoPP#1662",
        "heroName": "soldier76",
        "skillRating": 2400,
        "stats": {
            "hoursPlayed": 3,
            "wins": 11,
            "losses": 7,
            "gamesPlayed": 18,
            "kdRatio": 2.9936305732484074,
            "accuracy": 36,
            "blockedPerMin": 0,
            "healingPerMin": 148.0888888888889,
            "damagePerMin": 1338.638888888889,
            "avgObjElims": 10.5,
            "avgObjTime": 56.22222222222222,
            "pKdRatio": 1,
            "pAccuracy": 1,
            "pBlockedPerMin": 0,
            "pHealingPerMin": 1,
            "pDamagePerMin": 1,
            "pAvgObjElims": 1,
            "pAvgObjTime": 1
        }
    },
    {
        "platformDisplayName": "PwNShoPP#1662",
        "heroName": "hanzo",
        "skillRating": 2400,
        "stats": {
            "hoursPlayed": 3,
            "wins": 11,
            "losses": 7,
            "gamesPlayed": 18,
            "kdRatio": 2.9936305732484074,
            "accuracy": 36,
            "blockedPerMin": 0,
            "healingPerMin": 148.0888888888889,
            "damagePerMin": 1338.638888888889,
            "avgObjElims": 10.5,
            "avgObjTime": 56.22222222222222,
            "pKdRatio": 1,
            "pAccuracy": 1,
            "pBlockedPerMin": 0,
            "pHealingPerMin": 1,
            "pDamagePerMin": 1,
            "pAvgObjElims": 1,
            "pAvgObjTime": 1
        }
    },
    {
        "platformDisplayName": "PwNShoPP#1662",
        "heroName": "dva",
        "skillRating": 2400,
        "stats": {
            "hoursPlayed": 3,
            "wins": 11,
            "losses": 7,
            "gamesPlayed": 18,
            "kdRatio": 2.9936305732484074,
            "accuracy": 36,
            "blockedPerMin": 0,
            "healingPerMin": 148.0888888888889,
            "damagePerMin": 1338.638888888889,
            "avgObjElims": 10.5,
            "avgObjTime": 56.22222222222222,
            "pKdRatio": 1,
            "pAccuracy": 1,
            "pBlockedPerMin": 0,
            "pHealingPerMin": 1,
            "pDamagePerMin": 1,
            "pAvgObjElims": 1,
            "pAvgObjTime": 1
        }
    },
    {
        "platformDisplayName": "PwNShoPP#1662",
        "heroName": "zenyatta",
        "skillRating": 2400,
        "stats": {
            "hoursPlayed": 3,
            "wins": 11,
            "losses": 7,
            "gamesPlayed": 18,
            "kdRatio": 2.9936305732484074,
            "accuracy": 36,
            "blockedPerMin": 0,
            "healingPerMin": 148.0888888888889,
            "damagePerMin": 1338.638888888889,
            "avgObjElims": 10.5,
            "avgObjTime": 56.22222222222222,
            "pKdRatio": 1,
            "pAccuracy": 1,
            "pBlockedPerMin": 0,
            "pHealingPerMin": 1,
            "pDamagePerMin": 1,
            "pAvgObjElims": 1,
            "pAvgObjTime": 1
        }
    },
];

export default HEROES;
