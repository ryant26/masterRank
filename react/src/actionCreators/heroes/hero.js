import * as HeroActionTypes from 'actiontypes/hero';

export const addHero = (hero) => {
  return {
    type: HeroActionTypes.ADD_HERO,
    hero
  };
};

export const removeHero = (hero) => {
    return {
        type: HeroActionTypes.REMOVE_HERO,
        hero
    };
};

export const addHeroes = (heroes) => {
  return {
    type: HeroActionTypes.ADD_HEROES,
      heroes
  };
};

export const clearAllHeroes = () => {
  return {
    type: HeroActionTypes.CLEAR_ALL_HEROES
  };
};