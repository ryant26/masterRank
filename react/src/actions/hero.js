import * as HeroActionTypes from '../actiontypes/hero';

export const addHero = (hero) => {
  return {
    type: HeroActionTypes.ADD_HERO,
    hero
  }
}
