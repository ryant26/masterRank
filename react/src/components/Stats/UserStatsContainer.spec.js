import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import {getHeroes} from '../../resources/heroes';
import UserStatsContainer from './UserStatsContainer';
import {createStore} from "../../model/store";
import {addHero} from "../../actionCreators/hero";


const getMixedHeroes = () => {
    const heroes = getHeroes();
    heroes[0].platformDisplayName = 'notMyName';
    heroes[1].stats = undefined;
    return heroes;
};

const getStore = (heroes=getMixedHeroes()) => {
    let store = createStore();
    heroes.forEach((hero) => store.dispatch(addHero(hero)));
    return store;
};

describe('UserStatsContainer Component', () => {
    let wrapper;
    let hero;

    beforeEach(() => {
        const heroes = getMixedHeroes();
        hero = {
            platformDisplayName: heroes[1].platformDisplayName,
            skillRating: heroes[1].skillRating
        };
        wrapper = mount(
            <UserStatsContainer store={getStore()} hero={hero} invitable={false}/>
        );
    });

    it('should render without exploding', () => {
        let wrapper = mount(
            <UserStatsContainer store={getStore()} hero={hero} invitable={false}/>
        );

        expect(wrapper.find(UserStatsContainer)).toBeTruthy();
    });

    it('should match the snapshot', () => {
        let component = renderer.create(
            <UserStatsContainer store={getStore()} hero={hero} invitable={true}/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should have all the hero images for the users heroes', () => {
        let heroImages = wrapper.find('.HeroImages').children();

        expect(heroImages.length).toEqual(3);
    });

    it('should show the hero images in the order of preference', () => {
        let heroImages = wrapper.find('.HeroImages').children();
        let heroes = getHeroes();

        expect(heroImages.length).toBeGreaterThan(0);

        heroImages.forEach((image, i) => {
            let imageName = image.props().heroName;
            let heroName = heroes[i+1].heroName;

            expect(imageName).toEqual(heroName);
        });
    });
});