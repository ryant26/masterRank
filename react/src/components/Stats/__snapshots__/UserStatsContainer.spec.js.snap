// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserStatsContainer Component should match the snapshot 1`] = `
<div
  className="UserStatsContainer"
>
  <div
    className="header"
  >
    <div
      className="flex justify-between align-center"
    >
      <div>
        <h3>
          PwNShoPP#1662
        </h3>
        <span
          className="sub-title"
        >
          <b>
            2400
          </b>
           SR
        </span>
        <span
          className="sub-title"
        >
          <b>
            22
          </b>
           WINS
        </span>
      </div>
      <div
        className="HeroImages flex"
      >
        <div
          className=""
        >
          <img
            alt="hanzo icon"
            className="HeroImage"
            onClick={undefined}
            src={
              Object {
                "process": [Function],
              }
            }
          />
        </div>
        <div
          className=""
        >
          <img
            alt="dva icon"
            className="HeroImage"
            onClick={undefined}
            src={
              Object {
                "process": [Function],
              }
            }
          />
        </div>
        <div
          className=""
        >
          <img
            alt="zenyatta icon"
            className="HeroImage"
            onClick={undefined}
            src={
              Object {
                "process": [Function],
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="body"
  >
    <div
      className="HeroStatsList"
    >
      <div
        className="HeroStatsListItem"
      >
        <div
          className="flex align-center"
        >
          <div>
            <div
              className=""
            >
              <div
                className=""
              >
                <img
                  alt="hanzo icon"
                  className="HeroImage"
                  onClick={undefined}
                  src={
                    Object {
                      "process": [Function],
                    }
                  }
                />
              </div>
            </div>
          </div>
          <div>
            <h3>
              
              Hanzo
            </h3>
            <div
              className="sub-title"
            >
              Hero needs more games played
            </div>
          </div>
          <div
            className="flex justify-between record"
          >
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                N/A
              </div>
              <div
                className="sub-title"
              >
                WINS
              </div>
            </div>
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                N/A
              </div>
              <div
                className="sub-title"
              >
                LOSSES
              </div>
            </div>
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                N/A
              </div>
              <div
                className="sub-title"
              >
                K/D
              </div>
            </div>
          </div>
        </div>
        <div
          className="flex wrap justify-between"
        >
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              DAMAGE
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              HEALING
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              BLOCKED
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              OBJ. KILLS
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              OBJ. TIME
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              ACCURACY
            </div>
          </div>
        </div>
      </div>
      <div
        className="HeroStatsListItem"
      >
        <div
          className="flex align-center"
        >
          <div>
            <div
              className=""
            >
              <div
                className=""
              >
                <img
                  alt="dva icon"
                  className="HeroImage"
                  onClick={undefined}
                  src={
                    Object {
                      "process": [Function],
                    }
                  }
                />
              </div>
            </div>
          </div>
          <div>
            <h3>
              
              Dva
            </h3>
            <div
              className="sub-title"
            >
              3 hours played
            </div>
          </div>
          <div
            className="flex justify-between record"
          >
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                11
              </div>
              <div
                className="sub-title"
              >
                WINS
              </div>
            </div>
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                7
              </div>
              <div
                className="sub-title"
              >
                LOSSES
              </div>
            </div>
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                2.99
              </div>
              <div
                className="sub-title"
              >
                K/D
              </div>
            </div>
          </div>
        </div>
        <div
          className="flex wrap justify-between"
        >
          <div
            className="HeroStat flex flex-column"
          >
            <div>
              <span
                className="stat"
              >
                1339
              </span>
              <span
                className="stat-label sub-title"
              >
                /min
              </span>
            </div>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              DAMAGE
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <div>
              <span
                className="stat"
              >
                148
              </span>
              <span
                className="stat-label sub-title"
              >
                /min
              </span>
            </div>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              HEALING
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              BLOCKED
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <div>
              <span
                className="stat"
              >
                11
              </span>
              <span
                className="stat-label sub-title"
              >
                /min
              </span>
            </div>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              OBJ. KILLS
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <div>
              <span
                className="stat"
              >
                56
              </span>
              <span
                className="stat-label sub-title"
              >
                /min
              </span>
            </div>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              OBJ. TIME
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <div>
              <span
                className="stat"
              >
                36
              </span>
              <span
                className="stat-label sub-title"
              >
                %
              </span>
            </div>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              ACCURACY
            </div>
          </div>
        </div>
      </div>
      <div
        className="HeroStatsListItem"
      >
        <div
          className="flex align-center"
        >
          <div>
            <div
              className=""
            >
              <div
                className=""
              >
                <img
                  alt="zenyatta icon"
                  className="HeroImage"
                  onClick={undefined}
                  src={
                    Object {
                      "process": [Function],
                    }
                  }
                />
              </div>
            </div>
          </div>
          <div>
            <h3>
              
              Zenyatta
            </h3>
            <div
              className="sub-title"
            >
              3 hours played
            </div>
          </div>
          <div
            className="flex justify-between record"
          >
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                11
              </div>
              <div
                className="sub-title"
              >
                WINS
              </div>
            </div>
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                7
              </div>
              <div
                className="sub-title"
              >
                LOSSES
              </div>
            </div>
            <div
              className="RecordStat flex flex-column align-end"
            >
              <div
                className="stat"
              >
                2.99
              </div>
              <div
                className="sub-title"
              >
                K/D
              </div>
            </div>
          </div>
        </div>
        <div
          className="flex wrap justify-between"
        >
          <div
            className="HeroStat flex flex-column"
          >
            <div>
              <span
                className="stat"
              >
                1339
              </span>
              <span
                className="stat-label sub-title"
              >
                /min
              </span>
            </div>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              DAMAGE
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <div>
              <span
                className="stat"
              >
                148
              </span>
              <span
                className="stat-label sub-title"
              >
                /min
              </span>
            </div>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              HEALING
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <span
              className="stat"
            >
              -
            </span>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              BLOCKED
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <div>
              <span
                className="stat"
              >
                11
              </span>
              <span
                className="stat-label sub-title"
              >
                /min
              </span>
            </div>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              OBJ. KILLS
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <div>
              <span
                className="stat"
              >
                56
              </span>
              <span
                className="stat-label sub-title"
              >
                /min
              </span>
            </div>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              OBJ. TIME
            </div>
          </div>
          <div
            className="HeroStat flex flex-column"
          >
            <div>
              <span
                className="stat"
              >
                36
              </span>
              <span
                className="stat-label sub-title"
              >
                %
              </span>
            </div>
            <div
              className="indicator"
            >
              <div
                className="indicator-level"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </div>
            <div
              className="sub-title"
            >
              ACCURACY
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="footer flex justify-end align-center"
  >
    <div
      className="button-secondary flex align-center"
      onClick={[Function]}
    >
      <div
        className="button-content"
      >
        INVITE TO GROUP
      </div>
    </div>
  </div>
</div>
`;
